rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+```rust
+/// A persistent key-value pair mapping struct.
+pub struct StorageMap<K, V> {}
+
+impl<K, V> StorageMap<K, V> {
+    #[storage(write)]
+    pub fn insert(self: StorageHandler<Self>, key: K, value: V) {
+        let key = sha256((key, self.key));
+        store::<V>(key, value);
+    }
+
+    #[storage(read)]
+    pub fn get(self: StorageHandler<Self>, key: K) -> Option<V> {
+        let key = sha256((key, self.key));
+        get::<V>(key)
+    }
+
+    #[storage(write)]
+    pub fn remove(self: StorageHandler<Self>, key: K) -> bool {
+        let key = sha256((key, self.key));
+        clear::<V>(key)
+    }
+}
+```
+
+Note that this requires that the compiler allows `self` to be type-ascribed. This is not possible today but should be easy to do. We should make sure to restrict the types allowed to only `Self` and `StorageHandler<Self>`. Rust has a small list of types that `self` is allowed to have such as `Rc` and `Arc` so this feature is not unheard of.
+
+With the above new implementation, we should be able to continue to use `StorageMap` as before:
+
+```rust
+storage {
+    my_map: StorageMap<u64, u64> = StorageMap {}
+}
+
+storage.my_map.insert(0, 0);
+let x = storage.my_map.get(0);
+storage.my_map.remove(0);
+```
+
+A similar approach can be followed to re-implement `StorageVec` and other dynamic storage types.
+
+## Dynamic Storage Types in Structs
+
+The code below will now be valid after implementing all of the above:
+
+```rust
+MyStruct {
+    map1: StorageMap<u64, u64>,
+    map2: StorageMap<u64, u64>,
+}
+
+storage {
+    s: Mystruct = MyStruct {
+        map1: StorageMap { },
+        map2: StorageMap { },
+    }
+}
+
+storage.s.map1.insert(0, 0);
+let x = storage.s.map2.get(0);
+```
+
+It will also be possible to pass a `StorageMap` to a function by passing a `StorageHandler` to it as follows:
+
+```rust
+storage {
+    my_map: StorageMap<u64, u64> = StorageMap {}
+}
+
+fn foo(map: StorageHandler<StorageMap<u64, u64>>) {
+    map.insert(0, 0);
+}
+
+fn bar() {
+    foo(storage.my_map); 
+}
+```
+
+Sway code using storage maps and vectors will not look any different than before after this RFC is implemented.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+There are two main drawbacks for the approach proposed above:
+
+1. Reading and writing a storage variable is now a bit more complicated and less ergonomic than before because we now have to call `store` and `get` manually. We also have to handle the `Option` returned by `get` which can be quite verbose. We should consider making this more ergonomic by introducing a simpler syntax that de-sugars to the API calls above. For example, we could de-sugar something like `storage.x = 42` to `storage.x.store(42)`. We could also consider making `get` return the data directly instead of an `Option` if we completely disallow storage slots corresponding to storage variables from being invalidated (For example, remove `std::storage::clear`).
```

Comment:
While considering these alternatives, we should think about whether or not we really *want* to provide simpler syntax.

I think the user should at least feel the difference in cost of using storage. Hiding interaction behind e.g. an assignment operator, while *slightly* more concise, looks and feels too much like a regular in-memory stack assignment, and I worry new users will confuse the costs of the two?

Reply:
> We could also consider making `get` return the data directly instead of an `Option` if we completely disallow storage slots corresponding to storage variables from being invalidated (For example, remove `std::storage::clear`).

What is the current use-case for `clear`?

Reply:
I'm thinking to prevent sentinel values.  If a value is no longer valid then it should be `None`, not zero, which clearing effectively provides.

Reply:
> This is why I'm in favour of not requiring an initialiser either

This is reasonable to me, particularly because writing initializers today, especially for complex structs, is a pain in the behind.

Reply:
Keeping initializers for now for reasons related to packing structs.